### PR TITLE
enforce correct dtype promotion in qgt implementations

### PR DESCRIPTION
**Pull request by PhilipVinc**
numpy.sqrt() was not a weak type and was promoting float32 neetworks to float64.
This was an issue on `QGTJacobian***`

(This came up because I'm trying to run netket at single precision on gpu to get a speed bump and lower memory requirements).

@inailuig can I ask you to add a test for this?

### CODE REVIEW

File: netket/optimizer/qgt/qgt_jacobian_pytree_logic.py [hunk 1]
```diff
@@ -114,7 +114,8 @@ def _divide_by_sqrt_n_samp(oks, samples):
     divide Oⱼₖ by √n
     """
     n_samp = samples.shape[0] * mpi.n_nodes  # MPI
-    return jax.tree_map(lambda x: x / np.sqrt(n_samp, dtype=x.dtype), oks)
+    sqrt_n = float(np.sqrt(n_samp, dtype=x.dtype))  # enforce weak type
```
Comment (inailuig) [inline netket/optimizer/qgt/qgt_jacobian_pytree_logic.py]:
```suggestion
    sqrt_n = math.sqrt(n_samp)  # enforce weak type
```
(shorter but requires importing math, usually using this in my own code)

**Comment by inailuig** [comment]
Still it is puzzling to me how a float32 numpy array is suddenly promoted to float64 on the gpu ?

**Comment by PhilipVinc** [comment]
Where are you testing this?
On my local machine it works fine

File: netket/optimizer/qgt/qgt_jacobian_pytree_logic.py (hunk 1 above)
Comment (PhilipVinc) [reply netket/optimizer/qgt/qgt_jacobian_pytree_logic.py]: thanks! that's a good idea

**Comment by PhilipVinc** [comment]
maybe you're on izar and  forgot to update netket to this branch?

**Comment by inailuig** [comment]
> Where are you testing this?
> On my local machine it works fine
my bad, on the cluster I had an outdated version of netket (pre https://github.com/netket/netket/pull/1155)

@PhilipVinc So with this PR you want to use a different fix (using sth implicitly weak) than I did in that PR (setting the dtype), and we simply want a regression test to check the dtypes stay correct?

**Comment by PhilipVinc** [comment]
ah.
Mmmm...
Maybe there was no issue then?
But I was seeing an issue.

**Comment by PhilipVinc** [comment]
Ah, I think that in my own branch I was in a pre-#1155 world

**Pull request closed without merging**